### ISSUE TITLE
Ability to have localized registry/git cache

### ISSUE THREAD

**Issue by devsnek** [rust-lang/cargo#9021]
**Describe the problem you are trying to solve**

An open source project I'm working on uses a number of tools for building the final artifacts. One of the many tools used is cargo. Cargo is the only tool among them which forces cached data to be stored in a single global location. Many of the contributors are not programmers, and we'd like to avoid cluttering their computers with files they don't even know about.

**Describe the solution you'd like**

Ideally, cargo could have an option to put all the relevant cached objects in `target/.cargo` or something like that. It is worth mentioning that this should not affect how configuration resolution works (this is why `$CARGO_HOME` is undesirable).

**Notes**

I'm happy to code this, just need the cargo gods to come to a decision on whether the feature is wanted in the first place.

**Comment by Eh2406** [comment]
Would it make sense to use cargo vender for your use case?

**Comment by devsnek** [comment]
@Eh2406 i don't think so. it would be undesirable to hold all those files in version control.

**Comment by joshtriplett** [comment]
Would it suffice to have a separate environment variable to specify the cargo configuration file location? That way you could set `CARGO_HOME` without affecting configuration file resolution.

Another option: you could detect if the user already *has* a cargo directory, and only set `CARGO_HOME` if they don't.

**Comment by devsnek** [comment]
Maybe a separate env variable to set the git/crates cache location?

**Comment by joshtriplett** [comment]
@devsnek Well, hopefully one day those will go in `XDG_CACHE_HOME`, which you could set arbitrarily, but that's going to take a while.

**Comment by jonhoo** [comment]
As a quick note, this seems very related to https://github.com/rust-lang/cargo/issues/6452.

**Comment by epage** [comment]
If a separate environment variable is sufficient for this, then this is covered by #1734 and I'm going to close in favor of that.  If there is a reason for us to re-evaluate, let us now!
